modules/internal/String.chpl

```diff
@@ -1798,6 +1798,137 @@ module String {
+       lines. Empty lines are lines containing only whitespace. Tabs and spaces
+       are considered whitespace, but are not treated as the same characters
+       when determining common whitespace.
+
+       When ``columns > 0``, remove ``columns`` leading whitespace characters
+       from each line. Tabs are not considered whitespace when ``columns > 0``,
+       so only leading spaces are removed.
+
+       :arg columns: The number of columns of indentation to remove. Infer
+                     common leading whitespace if ``columns == 0``.
+       :type columns: `int`
+
+       :arg ignoreFirst: When ``true``, ignore first line when determining the
+                         common leading whitespace, and make no changes to the
+                         first line.
+       :type ignoreFirst: `bool`
+
+       :returns: A new `string` with indentation removed.
+
+       .. warning::
+
+          ``string.dedent`` is not considered stable and is subject to change in
+          future Chapel releases.
+    */
+    proc string.dedent(columns=0, ignoreFirst=true): string {
+      if chpl_warnUnstable then
+        compilerWarning("string.dedent is subject to change in the future.");
+
+      const low = if ignoreFirst then 1 else 0;
+      var lines = this.split('\n');
+      var ret = '';
+
+      if columns <= 0 {
+        // Find common leading whitespace across all non-empty lines
+        const margin = computeMargin(lines[low..]);
+
+        // Remove margins from all lines if it's not empty
+        if margin.size > 0 {
+          for line in lines[low..] {
+            // Compute offset
+            var offset = 0;
+            if !isWhitespaceOnly(line) {
+              offset = margin.size;
+            } else {
+              // Remove margin as long as it matches for empty lines
+              for i in 0..<min(margin.size, line.size) {
+                if line[i] != margin[i] then break;
+                offset += 1;
+              }
+            }
+            // Remove margin from line
+            line = line[offset..];
+          }
+        }
+      } else {
+        // Remove up to `columns` number of spaces from each line
+        for line in lines[low..] {
+          // Note: We only consider spaces (not tabs) for columns > 0
+          const indent = line.size - line.strip(' ', trailing=false).size;
+          const offset = min(indent, columns);
+          line = line[offset..];
+        }
+      }
+
+      ret = '\n'.join(lines);
+      return ret;
+    }
+
+    /* Compute margin of common leading white space across lines in a string.
+       Spaces and tabs are respected.
+     */
+    private proc computeMargin(lines: [] string): string {
+      var margin = '';
+
+      for line in lines {
+
+        // Skip empty lines
+        if isWhitespaceOnly(line) {
+          continue;
+        }
+
+        // Determine leading whitespace (spaces and tabs) in line
+        var indent = '';
```

Comment:
can we make this variable name (slightly) more verbose? `curMargin`, `curIndent` etc?